### ISSUE TITLE
Error when I try to install Openswoole

### ISSUE THREAD

**Issue by swadams2** [openswoole/ext-openswoole#359]
1. What did you do? If possible, provide a simple script for reproducing the error.
```
I tried to install Openswoole according to the instructions at openswoole.com.  It seemed to work until the last step, which is:
composer require openswoole/core:22.1.5

It yielded this error:
 Package openswoole/core at version 22.1.5 has requirements incompatible with your PHP version, PHP extensions and Composer version:
    - openswoole/core 22.1.5 requires ext-openswoole >=22.0 but it is not present.

I tried installing ext-openswoole according to the instructions on this site, but the above command still gave the same error.

I then tried "composer require ext-openswoole" and got this:
Could not find package ext-openswoole.

  Did you mean one of these?
      upscale/ext-openswoole
      upscale/ext-swoole

So then I tried "composer require upscale/ext-openswoole" and got this:
Package upscale/ext-openswoole has requirements incompatible with your PHP version, PHP extensions and Composer version:
    - upscale/ext-openswoole v22.1.2 requires ext-openswoole self.version but it is not present.
    
```
2. What did you expect to see?
```


```
3. What did you see instead?
```


```
4. What version of OpenSwoole are you using (show your `php --ri openswoole`)?
```


```
5. What is your machine environment used (show your `uname -a` & `php -v` & `gcc -v`) ?
```
uname -a:
Linux swadams-HP-EliteDesk-800-G3-DM-35W 6.5.0-35-generic #35~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue May  7 09:00:52 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

php -v:
PHP 8.3.7 (cli) (built: May 13 2024 15:38:09) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.3.7, Copyright (c) Zend Technologies
    with Zend OPcache v8.3.7, Copyright (c), by Zend Technologies

gcc -v:
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/11/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none:amdgcn-amdhsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 11.4.0-1ubuntu1~22.04' --with-bugurl=file:///usr/share/doc/gcc-11/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++,m2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-11 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-bootstrap --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --enable-default-pie --with-system-zlib --enable-libphobos-checking=release --with-target-system-zlib=auto --enable-objc-gc=auto --enable-multiarch --disable-werror --enable-cet --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none=/build/gcc-11-XeT9lY/gcc-11-11.4.0/debian/tmp-nvptx/usr,amdgcn-amdhsa=/build/gcc-11-XeT9lY/gcc-11-11.4.0/debian/tmp-gcn/usr --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu --with-build-config=bootstrap-lto-lean --enable-link-serialization=2
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.4.0 (Ubuntu 11.4.0-1ubuntu1~22.04)



```

You can also try the following OpenSwoole support channels:

* [Documentation](https://openswoole.com/docs) - Documentation for Open Swoole
* [Slack](https://goo.gl/forms/wooTTDmhbu30x4qC3) - Slack channel of Open Swoole
* [Discord](https://discord.gg/5QC57RNPpw) - Discord server of Open Swoole

**Comment by bitslip6** [comment]
You need to install the extension. composer only installs php packages (php code, not code compiled into your PHP install)

git clone https://github.com/openswoole/ext-openswoole
cd ext-openswoole
phpize
./configure
make 
make install
etc...

run "php -i" make sure that swoole is listed as an extension.

**Comment by swadams2** [comment]
I tried that and nothing changed.  Oh, and I used the "php -m" command and saw this error: 
PHP Warning:  PHP Startup: Unable to load dynamic library 'openswoole' (tried: /usr/lib/php/20230831/openswoole (/usr/lib/php/20230831/openswoole: cannot open shared object file: No such file or directory), /usr/lib/php/20230831/openswoole.so (/usr/lib/php/20230831/openswoole.so: cannot open shared object file: No such file or directory)) in Unknown on line 0

**Comment by bitslip6** [comment]
sounds like maybe your phpize doesnt match your PHP install.
run:
which php
which phpize

and make sure those point to the same php install.
also try:
find /usr/lib/php -name openswoole.so

**Comment by swadams2** [comment]
That may be the case.  Output of "ls -l /usr/lib/php" shows:
drwxr-xr-x 3 root root 4096 May 22 12:12 20210902
drwxr-xr-x 2 root root 4096 May 20 16:25 20230831
drwxr-xr-x 3 root root 4096 May 20 15:37 8.1
drwxr-xr-x 3 root root 4096 May 20 15:39 8.3
-rw-r--r-- 1 root root   13 Jan  9 06:01 packaging
-rwxr-xr-x 1 root root  514 Jan  9 06:01 php-fpm-socket-helper
-rw-r--r-- 1 root root 4845 Jan  9 06:01 php-helper
-rw-r--r-- 1 root root 9234 Jan  9 06:01 php-maintscript-helper
-rwxr-xr-x 1 root root 2976 Jan  9 06:01 sessionclean

I looked in the 20210902 directory and openswoole.so is there, but Composer looked in the 20230831 directory.

**Comment by swadams2** [comment]
I used the "which php" and "which phpize" commands and they yielded "/usr/bin/php" and "/usr/bin/phpize", respectively.

**Comment by dompie** [comment]
It has nothing to do with comopser. Composer only tells you, that your PHP needs the openswoole extension and it will not install the openswoole PHP classes, as long as the extension is missing.

I had the same (or similar) problem with openswoole on ubuntu somewhere early last year. I think I fixed it by making sure `phpize`, `php -i` and `php-config` have the same version output.
```
$ phpize --version
Configuring for:
PHP Api Version:         20220829
Zend Module Api No:      20220829
Zend Extension Api No:   420220829
```
You will find the versions in `php-config` and `php -i` output too. They all must match. Otherwise the extension will be build for another PHP version, or some other stuff happens that does not work.

Even when build succeeds, make sure openswoole is enabled on by checking if /etc/php/8.2/cli/mods-available/swoole.ini is linked to /etc/php/8.2/cli/conf.d/30-swoole.ini (or similar name). Sometimes this fails too.

**Comment by swadams2** [comment]
Oddly, in my /etc/php/8.3/cli/conf.d directory, I have 25-swoole.ini and 30-openswoole.ini, but 30-openswoole.ini links to a nonexistent file.

**Comment by swadams2** [comment]
I finally got it to work.  For some reason, there were still files related to PHP 8.1 in the /usr/include/php and /usr/lib/php directories even though I'm using 8.3 and this seems to have been the cause of the problem.  Using an "apt uninstall" command to get rid of 8.1 didn't work since apt said 8.1 wasn't installed.  I deleted them manually and reran the commands Bitslip6 listed.  I got a better error, one that said a conflicting extension had been loaded already.  I deleted the 25-swoole.ini link, restarted Apache, and it worked.  I was then able to get the composer command to work after some changes to file permissions/ownership.

**Comment by swadams2** [comment]
I finally got it to work. For some reason, there were still files related to PHP 8.1 in the /usr/include/php and /usr/lib/php directories even though I'm using 8.3 and this seems to have been the cause of the problem. Using an "apt uninstall" command to get rid of 8.1 didn't work since apt said 8.1 wasn't installed. I deleted them manually and reran the commands Bitslip6 listed. I got a better error, one that said a conflicting extension had been loaded already. I deleted the 25-swoole.ini link, restarted Apache, and it worked. I was then able to get the composer command to work after some changes to file permissions/ownership.
